compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -0,0 +1,1418 @@
+package dotty.tools.dotc
+package transform
+package init
+
+import core.*
+import Contexts.*
+import Symbols.*
+import Types.*
+import StdNames.*
+import NameKinds.OuterSelectName
+import NameKinds.SuperAccessorName
+
+import ast.tpd.*
+import util.SourcePosition
+import config.Printers.init as printer
+import reporting.StoreReporter
+import reporting.trace as log
+
+import Errors.*
+import Trace.*
+import Util.*
+
+import scala.collection.immutable.ListSet
+import scala.collection.mutable
+import scala.annotation.tailrec
+import scala.annotation.constructorOnly
+
+/** Check initialization safety of static objects
+ *
+ *  The problem is illustrated by the example below:
+ *
+ *      class Foo(val opposite: Foo)
+ *      case object A extends Foo(B)     // A -> B
+ *      case object B extends Foo(A)     // B -> A
+ *
+ *  In the code above, the initialization of object `A` depends on `B` and vice versa. There is no
+ *  correct way to initialize the code above. The current checker issues a warning for the code
+ *  above.
+ *
+ *  At the high-level, the analysis has the following characteristics:
+ *
+ *  1. The check enforces the principle of "initialization-time irrelevance", which means that the
+ *     time when an object is initialized should not change program semantics. For that purpose, it
+ *     enforces the following rule:
+ *
+ *         The initialization of a static object should not directly or indirectly read or write
+ *         mutable state of another static object.
+ *
+ *     This principle not only put initialization of static objects on a solid foundation, but also
+ *     avoids whole-program analysis.
+ *
+ *  2. The design is based on the concept of "cold aliasing" --- a cold alias may not be actively
+ *     used during initialization, i.e., it's forbidden to call methods or access fields of a cold
+ *     alias. Method arguments are cold aliases by default unless specified to be sensitive. Method
+ *     parameters captured in lambdas or inner classes are always cold aliases.
+ *
+ *  3. It is inter-procedural and flow-sensitive.
+ *
+ *  4. It is object-sensitive by default and parameter-sensitive on-demand.
+ *
+ *  5. The check is modular in the sense that each object is checked separately and there is no
+ *     whole-program analysis. However, the check is not modular in terms of project boundaries.
+ *
+ */
+object Objects:
+
+  // ----------------------------- abstract domain -----------------------------
+
+  sealed abstract class Value:
+    def show(using Context): String
+
+
+  /**
+   * A reference caches the values for outers and immutable fields.
+   */
+  sealed abstract class Ref(
+    valsMap: mutable.Map[Symbol, Value],
+    varsMap: mutable.Map[Symbol, Heap.Addr],
+    outersMap: mutable.Map[ClassSymbol, Value])
+  extends Value:
```

Comment:
`Ref` is a parent class for both `ObjectRef` and `OfClass`, which correspond to Object and Instance locations as listed in Section 4.4. Both `ObjectRef` and `OfClass` need to store the values of immutable fields, as stored in `valsMap`, and the abstract heap address for mutable fields, as stored in `varsMap`, and the pointers and values for outer classes, as listed in `outersMap`.